### DISCUSSION TITLE
Boundary effects

### DISCUSSION THREAD

**Discussion by OverLordGoldDragon** [kymatio/kymatio#752] [General]
Zero padding yields severe energy deficit for large scale wavelets. Using corrected coefficients as described [here](https://github.com/kymatio/kymatio/discussions/732#discussioncomment-855702) (except `S1` is still `U1 * phi_t`), I obtain for WGN:

```
reflect      | zero
.039 (.001%) | .002 (.000%) -- S0
1749 (59.0%) | 108  (29.3%) -- S1
562  (19.0%) | 38.1 (10.2%) -- phi_t * phi_f
544  (18.4%) | 32.4 (8.73%) -- phi_t * psi_f
2.60 (.087%) | 142  (38.3%) -- psi_t * phi_f
50.7 (1.71%) | 25.1 (6.77%) -- psi_t * psi_f_up
50.7 (1.71%) | 24.2 (6.54%) -- psi_t * psi_f_down
1.381        | 0.173        -- E_out / E_in
# 1.381 > 1 because `S1` is duplication, kept for reference
```

This propped up when I could not explain why `phi_t * phi_f` and `phi_t * psi_f` energies did not sum to `S1`'s. The problem is much worse in frequential scattering per small `shape_fr` where we have wavelets whose time support can be **50x** greater; only a small fraction of the wavelet ever lands on non-zero data. 

`recalibrate` remedies this (and so does `exclude`) to a fair extent; `recalibrate` + `reflect`:

```
.039 (.001%) -- S0
1749 (48.0%) -- S1
881  (24.2%) -- phi_t * phi_f
695  (19.1%) -- phi_t * psi_f
12.2 (.337%) -- psi_t * phi_f
149  (4.11%) -- psi_t * psi_f_up
149  (4.11%) -- psi_t * psi_f_down
1.697        -- E_out / E_in
```
Note `phi_t * phi_f` and `phi_t * psi_f` nearly sum to `S1` (1576 vs 1749). But it's still off because of frequential zero padding. 

The questions are:

 1. Can we non-zero pad in time? If so, how?
 2. Can we non-zero pad in frequency? If so, how?

<hr>

### 1. Temporal padding

I figure an affirmative: `reflect` works. With 'corrected' coefficients, `_joint_lowpass()` no longer draws from reflected spins. Though it's still desired to have concat-able coeffs without exploding size, so I'm not yet convinced on doing nothing after `psi_t * psi_f`... leaving unanswered for now.

In padding time, one goal is complementing energy: with zero padding, exactly at boundary, half of every wavelet operates on nothing, and either loses energy or gains it (as it might otherwise have correlated to zero), both artificially. The idea's to provide the best guess of signal's continuation - a "prior". 

<hr>

### 2. Frequential padding

What's such a guess for frequential padding? I argue, `conj-reflect`. We can't:

  - Go past Nyquist, else we falsely raise the rate of information. 
  - Extend all in-between, akin to increasing `Q_fr`. But the effect is stretching, or interpolating within existing bounds, rather than extending past bounds. It also requires recalibrating wavelets to preserve physical dimensionality ... back to square one.

What'd happen in continuous time? No "Nyquist", simply insert outputs of higher `xi` wavelets. ... but we can do that. Just like with DFT, we'd _reflect_. So it's consistent with sampling the correct, continuous-time operation, with aliasing.

Except we'd not reflect; we'd reflect _with conjugation_. This conveniently takes care of spin: a simple reflection would be akin to padding with opposite spin, which'd yield cancellation. Instead it's padding with anti-analytic `psi_t` and same spinned `psi_f`. This not only complements energy but amplifies FDTS discriminability, both for and against.

An important caveat is ... it'd be `conj-reflect` + `zero` padding. Former about higher frequencies, latter about lower, per the [`j2 > j1` idea](https://github.com/kymatio/kymatio/pull/674#issuecomment-814815014). We prioritize zeros, ensuring they pad at most by `padded_len // 2` (more is never needed for either side), but also by not more than `shape_fr_max - len(coeff_list)`, as zeros are placeholders for the "missing coeffs".

I'm currently testing these ideas.

**Comment by OverLordGoldDragon** [comment]
I overlooked that we still have `phi_t * phi_f` over spinned pairs, which rules out `reflect` for temporal. The sole source of error here is `phi_t` drawing opposite spin from the reflection. Our main goal is to complement energy. Then, we could conjugate the reflection before convolving with spins, which'd avert opposite spin at reflection while complementing energy.

**Reply by OverLordGoldDragon** [reply]
It works:

![image](https://user-images.githubusercontent.com/16495490/121788801-53dcee00-cbe1-11eb-9058-3cb3e7eacc1c.png)

`E_down / E_up = 100`, down from previous `215` (but we've already established the metric's flawed). Still can't get `phi_t * phi_f` and `phi_t * psi_f` to sum to `S1`, looking into.

**Comment by OverLordGoldDragon** [comment]
## Conj-reflect-zeros padding explanation

Plotting padding for up-echirp; `Y_2_arr.imag`, `n2==6`, `conj-reflect-zeros`:

<img src="https://user-images.githubusercontent.com/16495490/121794558-4db73380-cc1a-11eb-8154-1b1ef7100046.png" width="500">

![image](https://user-images.githubusercontent.com/16495490/121795378-26b03000-cc21-11eb-8dbe-206a8db74e29.png)

"blue, red" means first negative then positive imaginary peak, which resonates with the up spin / anti-analytic wavelet (so we convolution resonates with down spin, which flips the kernel and thus correlates with up).

Note without `conj` there's "red, blue", which'll resonate with opposite spin and kill discriminability. Also note sometimes _not_ conjugating is correct: we conjugate every _other_ reflection.


### Stop condition

Scheme is best thought of left-right padding, but here it's implemented in circular manner via right-padding. In frequency, we keep reflecting from right and left, but since `shape_fr_max` can be non-dyadic, they won't form a perfect loop. However, from right, we don't care what happens past `right_bound + padded_len / 2` (it's beyond wavelet's width by design), likewise `left_bound - padded_len / 2`. Thus we keep reflecting in each direction independently for equal lengths.

Resulting boundary for running example:

<img src="https://user-images.githubusercontent.com/16495490/121795002-00d55c00-cc1e-11eb-9c20-47d0ec1adae8.png" width="410">


### Conv example

Zooming on a frequential slice:

![image](https://user-images.githubusercontent.com/16495490/121795557-02555300-cc23-11eb-8e9e-9739eef67f8a.png)

**Reply by lostanlen** [reply]
Thanks @OverLordGoldDragon for the excellent sleuthing as always

> The problem is much worse in frequential scattering per small shape_fr where we have wavelets whose time support can be 50x greater; only a small fraction of the wavelet ever lands on non-zero data.

What's the use case for it? Should we really allow it?
I guess it goes back to our conversation about high `j_fr` vs low `j2`
Is that what you call `exclude` ?


> I overlooked that we still have phi_t * phi_f over spinned pairs, which rules out reflect for temporal. 

I agree.
Personally, i wouldn't worry too much about temporal padding. Practical applications (audio, biomedical, etc.) have signal lengths `N` far greater than `T` and `2**J`. So i'd expect that the user will divide their signal into constant-sized "chunks", potentially with overlap, and apply a window to taper off boundary effects: Hann, Tukey, Gaussian, etc.
BTW, i'd like to look into the possibility of integrating that directly into Kymatio so that `N` becomes an optional argument (only `J` will be required).

Of course, the issue with frequential padding remains, because it happens in between scattering orders and can't be anticipated (not at a huge computational cost at least). 

>  This not only complements energy but amplifies FDTS discriminability, both for and against.

I like this idea very much

**Reply by OverLordGoldDragon** [reply]
> I agree.

> the issue with frequential padding remains

Unsure we're on same page; the padding problem is solved, in terms of correctness.

It is, however, computationally intractable. Am working on that.

> What's the use case for it? Should we really allow it?

The use case is avoiding boundary effects and ensuring correctness. It's controlled by `max_pad_factor` and `max_pad_factor_fr` parameters, so optional. `exclude` is [this](https://github.com/kymatio/kymatio/discussions/721#discussioncomment-749458).

**Reply by lostanlen** [reply]
Right, i saw that later. Excellent job
What i was trying to say was "we can't reasonably ask the user to take care of doing frequential padding as they want, we need to provide a built-in solution for that"


> The use case is avoiding boundary effects and ensuring correctness

Let me rephrase: what's the use case of analyzing a signal or bounded support N by wavelets of scale >>N ?
This arises for high `j_fr` and low `j2`
Perhaps one solution is to exclude the those psi_f's. This will create a gap in the low-frequency region of the LP sum, but we might be able to compensate for it by locally reducing the scale `F` of the phi_f, which will be equivalent to reducing its bandwidth.

Something like: (0's and 1's roughly denote in-band and out-of-band frequency regions)

Before
```
phi  1 1 00 0000 000000000
psi  0 0 11 0000 000000000 (j=3)
psi  0 0 00 1111 000000000 (j=2)
psi  0 0 00 0000 111111111 (j=1)
```

After
```
phi  1 1 11 1111 000000000 (note the enlarged bandwidth)
(j=2 and j=3 excluded!)
psi 0 0 00 0000 11111111 (j=1)
```

This corresponds to using a different `J_fr` for different values of `j2`. `F` (the amount of final averaging) would stay the same

Is that a bad idea?

**Reply by OverLordGoldDragon** [reply]
> scale >>N

If "scale" means `J` or `T`, no use case; they're bounded `<= N` (likewise for `F`). We pad `>> N` not for scale `>> N` but to avoid boundary effects. Here's lowpass for `T = N = 2048`:

<img src="https://user-images.githubusercontent.com/16495490/125550839-5126ab59-0351-4a59-949f-90cb7fef27b5.png" width="500">


For some scale `J` the required padding is `16 * J`.

> Is that a bad idea?

It's [`exclude`](https://github.com/kymatio/kymatio/discussions/721#discussioncomment-749458), is it not?

> "we can't reasonably ask the user to take care of doing frequential padding as they want, we need to provide a built-in solution for that

Unsure what this means. If they want correctness, `max_pad_factor=None`. Speed, `=1`. `exclude` and `recalibrate` reduce the amount of required padding.

> Personally, i wouldn't worry too much about temporal padding.

I would. It's not just a practical but a fundamental problem: what to do with missing information? We cannot extend an echirp at Nyquist. As for practice, having x16-fold extension of input isn't a safe bet. The boundary is always reached, and a complete and correct implementation should handle it. (And of course, the solution can be reused for frequential padding, which is unaddressed by data extension).

**Reply by OverLordGoldDragon** [reply]
Unfortunately this padding breaks alignment... 

![image](https://user-images.githubusercontent.com/16495490/121993321-53e11780-cdb4-11eb-8819-8547af0290c6.png)


Time scattering shares same fate, energy peaks will shift toward boundaries, a form of distortion - but it matters more in freq scattering. Guess I should try to renormalize the wavelets to compensate for energy loss from zeros.

**Reply by OverLordGoldDragon** [reply]
But same happens with a multiple A.M. signal, with peaks now _inside_ the bounds. So it's fundamentally impossible to preserve alignment in the general case, as correlated peak location will vary with filter support, with narrowest being most accurate. If the lowpass is sufficiently wide, it'll merge all peaks, making every psi inaccurate, but ironically guaranteeing alignment.

So `reflect` isn't committing a special crime. Similar holds of regular, unconjugated `reflect` that kills echirp discriminability: the echirp could rise and fall within the signal itself, and temporal averaging would blend them for both spins. Then the true villain is lowpassing.

I figure either padding will prove useful in different cases, so I'll look into renorm.

**Reply by OverLordGoldDragon** [reply]
Unfortunately no such renorm exists, at least as a simple fixed rescaling: zero padding can both amplify and reduce correlations past boundaries. One must either live with energy distortions or use wavelet + input specific renormalization.

**Reply by lostanlen** [reply]
just chiming in with the friendly reminder that the design priority in my opinion is to ensure energy conservation for different signals passing through the same architecture; whereas ensuring energy conservation for the same signal passing through different architectures is secondary.

mathematicaly speaking.
(1) `(E(x) = E(y))  ==> (E(Sx) = E(Sy))` where `x` and `y` are two zero-mean signals is the priority
(2) `E(Sx) = E(Rx)` where `S` and `R` are different flavors of the scattering transform is less of a priority

not sure if your corrective terms aim at (1) or (2) or both?

**Reply by OverLordGoldDragon** [reply]
Neither; I seek (3) `E(Sx) = E(x)`, and only for the "standard" variants.

**Reply by OverLordGoldDragon** [reply]
I doubt we can guarantee (1) without (3), (3) ensures nothing is lost and nothing over-accounted, so if (3) fails we can come up with some `y` for which (1) fails.

**Reply by lostanlen** [reply]
(3) for all `S` and for all `x` implies both (1) and (2) which is ideal
but it's not feasible in practice (i doubt it is true for `Scattering1D`, `max_order=1`, `J` large, and `average=True`, to begin with) so i would suggest we fight the battles we can fight?

**Reply by OverLordGoldDragon** [reply]
I'm saying (1) can't happen without (3), and (3) for one variant doesn't warrant it for others so (2) is separate.

By `E(Sx) = E(x)` I mean _if_ we had an infinite depth scattering network. This can instead be verified by looking at `U`, which does not require infinite depth, and I've done this and confirmed (3) holds [here](https://github.com/kymatio/kymatio/discussions/732#discussioncomment-864097). Problem is, it's far from holding with zero padding.

I've concluded that deviations from (3) are entirely due to padding and imperfect Littlewood-Paley. I now sought to normalize zero padding to also satisfy (3) but there's nothing such we can pre-compute independent of `x`.

**Reply by lostanlen** [reply]
(1) can happen without (3) if you imagine that the `E(Sx)/E(x)` ratio could vary depending on `S` but will stay constant depending on `x`.

Thanks for your analysis! I agree that padding and LP make things more subtle, and is not covered by theory.

Besides the question of boundary, what i would see as vital is `E(Sx)=E(S x_tau)` where `tau` is a small translation or deformation of `x`, and `x` is zero near the boundaries.

**Reply by OverLordGoldDragon** [reply]
> (1) can happen without (3) if you imagine that the E(Sx)/E(x) ratio could vary depending on

Indeed, making it exactly `E(Sx) = E(x)` wasn't easy. (1) doesn't guarantee that _relative_ energies (of same transform's coeffs) are correct.

> E(Sx)=E(S x_tau)

The beauty of (3) is, every conceivable energy test will pass.

**Reply by OverLordGoldDragon** [reply]
Zero-pad renorm idea: upscale such that, with time-domain wavelet that's inside bounds, and centered between bounds, its energy in that portion equals its original full energy. This doesn't remedy the inflations or deflations due to incomplete inner products, but might work well for a general input.

**Comment by OverLordGoldDragon** [comment]
In conjugating reflections one must account for 1) not conjugating _at_ boundary; 2) stride. Latter's tricky with various odd-even interactions; instead of deriving an exact complicated expression, I've made a simple algorithm to compute exact boundary indices.

First a demo of 1:

```python
[..., 1, 2, 3, 4, 5, 6, 7, 8, 7, 6, 5, 4, 3, 2, 1, 2, 3, 4, 5, 6, 7, 8]
[..., 1, 2, 3, 4, 5, 6, 7, 8,-7,-6,-5,-4,-3,-2, 1, 2, 3, 4, 5, 6, 7, 8]

[..., 1, 2, 3, 4, 5, 6, 7, 6, 5, 4, 3, 2, 1, 2, 3, 4, 5, 6, 7]
[..., 1, 2, 3, 4, 5, 6, 7,-6,-5,-4,-3,-2, 1, 2, 3, 4, 5, 6, 7]
```

Now with stride:

```python
[..., 1, 3, 5, 7, 7, 5, 3, 1, 3, 5, 7]
[..., 1, 3, 5, 7,-7,-5,-3, 1, 3, 5, 7]

[..., 1, 5, 7, 3, 3, 7]
[..., 1, 5,-7,-3, 3, 7]


[..., 1, 3, 5, 7, 5, 3, 1, 3, 5, 7]
[..., 1, 3, 5, 7,-5,-3, 1, 3, 5, 7]

[..., 1, 5, 5, 1, 5]
[..., 1, 5,-5, 1, 5]
```
To determine exact border indices, we simulate a reflected ramp at the _original_ (unsubsampled) length, find its border indices via `diff`, then map to subsampled:

![image](https://user-images.githubusercontent.com/16495490/124700607-0ca2fe00-defe-11eb-93b0-7dcefa3f92b3.png)

It might seem that in `subsampled=4` case the leftmost boundary should move one sample to the right, but it's correct where it is:

```python
[..., 42, 43, 44, 45, 46, 47, 46, 45, 44, 43, 42, ...]
[..., 40, 44, 46, 42, ...]
```
The goal is `op(x[::k]) == op(x)[::k]`, which this accomplishes.

**Reply by OverLordGoldDragon** [reply]
Above was the initial variant, which complicated via requiring correction steps per directional non-uniqueness. The simpler and definitive approach, based on it, is:

 1. Simulate up ramp at input length `N`
 2. Pad to maximum, `2**J_pad`
 3. Label every up ramp with `+1`, including endpoints, a total of `N` points per ramp - and everything else `-1`
 4. Subsample resulting sequence
 5. Conjugate all `-1`s

Tricky is step 3, with `diff` being forward difference, will label left endpoint as `-1`, and is corrected via `ramp_diff = diff(ramp); ramp_diff[where(diff(ramp_diff)) > 0] = 1`.

**Comment by OverLordGoldDragon** [comment]
### `conj-reflect-zero` caveat

In the slice demo, note although **blue->red** is dominant within its own locality:

<img src="https://user-images.githubusercontent.com/16495490/127736343-c412a91a-e9d6-49eb-b6f1-acf85e3e5775.png" width="400">

a larger support wavelet that spans over the boundary will have a dominant **red->blue**. Note the requirement for such a support diminishes as we near Nyquist: it's **blue->red->blue->red**, which includes **red->blue**.

Latter case is fixed via a zero-pad "buffer" at boundary, or by ensuring no FDTS near Nyquist. Former is fixed by limiting `J_fr` and `F`. I confirmed both increase FDTS discriminability per L2 measure.

Depending on prevalence of these factors, `conj-reflect-zero` is **detrimental** to FDTS discriminability per L2 measure; `zero` is superior. However, `conj-reflect` *along time* is still beneficial (assuming sufficiently small `J, T`).

Regardless I still favor `conj-reflect-zero` as 1) L2 is problematic as described several times (here mainly inflating numbers via small changes in denominator), and `conj-reflect-zero` is **superior** or close-to if comparing on per-coefficient basis; 2) `zero` wrecks high-support wavelet energies and creates false correlations at boundaries.

**Reply by OverLordGoldDragon** [reply]
After a more comprehensive look, I conclude that `conj-reflect-zero` can meaningfully harm FDTS discriminability; however, it's still the favorable scheme for realistic inputs.

`conj-reflect-zero` creates false resonance with opposite spin as described, _and_ exacerbates false resonances that `zero` possesses, via `reflect`'s energy doubling. It also, however, eliminates false resonance due to incomplete correlation with large support wavelets, and strengthens matched-spin resonance.

Below is a description of exact sources of false resonance, comparing `zero` and `conj-reflect-zero`. But first an overview of how spin "resonates" maximally and into zero.

### Spin resonance

`psi_t` demodulates temporal variations, then `psi_f` resonates with FDTS; we look at latter, separably. Correlate analytic wavelet with itself and anti-analytic counterpart; former first:

![image](https://user-images.githubusercontent.com/16495490/127785278-ad14334f-7997-4349-b02f-7c34cb9869f6.png)

Morlet times Morlet is another Morlet with half original's temporal support and double the frequency - still zero-mean, and hence zero-sum (and zero correlation).

The zero sum is reproduced with any input that shares Morlet's symmetries: even in real, odd in imaginary, and imaginary positive peak side matching Morlet's (in this case, left).

Now for max resonance:

![image](https://user-images.githubusercontent.com/16495490/127785282-68447b7b-c8ae-4e2b-9303-c37a535142f4.png)


From freq-domain conv perspective, self-convolution is cross-correlation with own conjugate, i.e. squaring freq-domain waveform; above shows this for when the two are perfectly aligned, yielding peak correlation. Convolving with opposite spin, on the other hand, is freq-domain multiplication of a ~Gaussian with same ~Gaussian but on opposite frequency axis, yielding zero - as was confirmed in previous example.

Max resonance is produced from same symmetries except imaginary positive peak side is located opposite to Morlet's.

Note also that we can zero-pad the center of these wavelets; it'll only diminish norms of respective real and imaginary parts, but not change whether sum is zero or non-zero: this is same as convolving the wavelet with red->blue separated by boundary.

### Resonance criteria

Precise conditions are explored [here](https://dsp.stackexchange.com/q/76560/50076); one criterion is for ratio of L2 norms of input's and wavelet's cross-complex parts (one's real to other's imaginary, and vice versa) be equal. This is highly unlikely.

### Boundary spikes

Created by 1) discontinuity (e.g. zero padding); 2) `reflect`. 1 is simple false correlation with wavelet positioned at boundary; 2 is since `sum(real even * imag odd) = 0`. We have:

![image](https://user-images.githubusercontent.com/16495490/127785997-9de7cb15-d529-44cb-a192-090a11396f73.png)

Note the white vertical lines and corresponding large freq conv heatmaps; the real part is non-zero there, which manifests as a large correlation with `psi_f`.

### Transition spikes

Similar to boundary spikes but occur prior to FDTS finishing, where a vertical slice is incomplete (e.g. only blue->red rather than red->blue->red->blue):

![image](https://user-images.githubusercontent.com/16495490/128008650-900b48fc-85f5-44a9-8365-3ed655ea0ecd.png)

### Analyticity

A not strictly analytic frequential bandpass will draw from opposite spin. Additionally, not strictly analytic _temporal_ bandpass creates a real-imag L2 norm disparity for `U1 * psi_t` slices: 

![image](https://user-images.githubusercontent.com/16495490/128010635-902869fa-b8d9-40b0-854b-bd58d80be6d4.png)

The second to last waveform one slice of second-order time scattering; the dominant imaginary part persists across several such slices and creates norm disparity in frequential scattering --> false resonance.

### Fundamental limitation

An important conclusion is: **perfect JTFS is impossible** in a finite implementation; there's no way to rid of the "boundary spike" without breaking other things. 

  - Second-order temporal scattering takes a positive real sequence as input; when reflected, at boundary the imaginary part of the convolution will be zero (`sum(real even * imag odd)`). 
  - To avoid this we can negate the reflected sequence, but that makes the input negative which is antithetical to higher-order scattering. 
  - Other schemes will create false correlations similar to zero padding; only reflect-like can create an idealized continuum

I still imagine it possible to force a perfect JTFS for idealized input: extend echirp's frequency sweep, then unpad more in time _and_ frequency, which may eliminate all artifacts (except imperfect analyticity) and yield an infinite energy ratio. I've pushed the per-slice ratio beyond 1000,000 this way -- but it's meaningless hackery.

### Summary: False resonance

 1. **Large support** correlating red->blue across boundaries
 2. **Boundary spikes** per discontinuity / reflect
 3. **Transition spikes** per discontinuity; also occurs for imperfect FDTS
 4. **Norm disparity** between input's and wavelets' real/imag parts
 5. **Analyticity** for Morlet isn't strict

### 'Realistic' inputs

Anything that's not the ideal considered here will qualify; so long as the input isn't perfectly spin up or spin down, there will be both spins, and the effect of these imperfections easily gets drowned out. Simply adding two AM cosines already makes `conj-reflect-zero` superior to `zero`.

**Reply by OverLordGoldDragon** [reply]
Interesting to observe, the described time-domain assymetries of non-analyticity manifest as wiggles in the 2D wavelet's main lobe - `analytic=False` vs `True`:

<img src="https://user-images.githubusercontent.com/16495490/161472282-d31ebc7e-6522-4c3c-89fc-e9ab83a83ecd.gif" width="300">

a more prominent case

<img src="https://user-images.githubusercontent.com/16495490/167259779-07ec209d-35ef-487d-ad58-18b121110847.gif" width="100">


While `True` doesn't eliminate curves, it only leaves some toward edges where their effect is much weaker. The bottom right wavelet is both fully analytic and untampered, and is perfectly straight, as are all the wavelets here

<img src="https://user-images.githubusercontent.com/16495490/161472621-feb66806-2fc9-4c34-b38f-d7badedb75e2.png" width="300">

It's interesting that perfect analyticity doesn't require perfect straight, but not that I studied analyticity in 2D.
